harness/determined/experimental/inference/_torch_batch_process.py

```diff
@@ -0,0 +1,405 @@
+
+        slots_per_node = len(info.slot_ids)
+        num_nodes = len(info.container_addrs)
+        total_worker = num_nodes * slots_per_node
+        # Get global rank
+        rank = core_context.distributed.get_rank()
+        latest_checkpoint = info.latest_checkpoint
+        skip = 0
+
+        # Synchronize default output uuid
+        if rank == 0:
+            default_output_uuid = str(uuid.uuid4())
+            core_context.distributed.broadcast(default_output_uuid)
+        else:
+            default_output_uuid = core_context.distributed.broadcast(None)
+
+        # Get previous trial state from checkpoint if available
+        if latest_checkpoint is not None:
+            logging.info("Checkpoint is not none")
+            with core_context.checkpoint.restore_path(latest_checkpoint) as path:
+                metadata = _load_state(path)
+                skip = metadata["steps_completed"]
+                logging.info(f"Previous run completed {skip} steps")
+                default_output_uuid = metadata["default_output_uuid"]
+
+        output_uuid_with_rank = default_output_uuid + f"/rank_{rank}"
+
+        batch_processor_context = TorchBatchProcessorContext(core_context, output_uuid_with_rank)
+
+        per_batch_processor = batch_processor_cls(
+            context=batch_processor_context,
+        )
+
+        dataloader = DataLoader(
+            dataset=dataset, batch_size=batch_size, shuffle=False, **dataloader_kwargs
+        ).get_data_loader(repeat=False, skip=skip, num_replicas=total_worker, rank=rank)
+
+        # Create dummy searcher op to report progress to master
+        dummy_searcher_op = None
+        # Initialize dummy searcher for progress report
+        if rank == 0:
+            dummy_searcher_op = core.DummySearcherOperation(1, True)
+
+        dataloader_iterator = iter(dataloader)
+
+        # Enumerate over dataloader directly may cause some workers to iterate for 1 more time
+        # than others when drop_last = False. If those workers synchronize on the last batch_idx,
+        # they would hang forever as other workers never hit that last batch_idx.
+        # To avoid the issue, we calculate and take the ceiling of the iteration count to ensure
+        # all workers iterate for the same number of times.
+        max_batch = math.ceil(dataset_len / batch_size / total_worker)
+        iterate_length = _validate_iterate_length(iterate_length, max_batch)
+
+        last_checkpoint_idx = 0
+        batch_idx = skip
+        steps_completed = skip
+
+        """
+        (2) Run batch processing
+        """
+
+        for batch_idx in range(skip, iterate_length):
+            X = next(dataloader_iterator, None)
+            if X is not None:
+                per_batch_processor.process_batch(batch=X, batch_idx=batch_idx)
+            steps_completed = batch_idx + 1
+
+            # Checkpoint and check preemption
+            if (batch_idx + 1) % checkpoint_interval == 0:
```

Comment:
you probably need a check somewhere to handle `checkpoint_interval = 0` case

Reply:
Yes, great point!